#  ```pyoti``` library Readme
### Lite version
# V 1.2
```python
""" 
Contributors to this file:
Mauricio Aristizabal Cano

Date of creation:  20 04 2016
Last modification: 27 03 2023
Universidad EAFIT, Medellin, Colombia.
University of Texas at San Antonio
"""
```

## 0. Introduction

The main purpose of this document is to explain the capabilities of the sparse OTI library for python 3: ```pyoti``` . As a basis for the explanation, several examples will be given in order to show and understand the capabilities of the library, as well as its basic advantages.

As a result, there will be several justifications based on an "a priori" utilisation of the library.


## 1. Getting started
In order to import the libary, one way of doing so is to call it with the following command:

In [1]:
import pyoti.sparse as oti  # Import the library. 'oti' stands for 'Order Truncated Imaginary number'

### 1.1 Basics
#### 1.1.1 OTI number creation

In order to create an OTI number, one manner to do so is as follows

In [46]:
a = 10 + oti.e(1) + 5.2* oti.e([3,4]) + 3*oti.e([[2,3],4])
print(a)

10 + 1 * e([1]) + 5.2 * e([3,4]) + 3 * e([[2,3],4])


In [48]:
a = 10 + oti.e(1) + 5.2* oti.e([3,4]) + 3*oti.e([2,2,2,4])
print(a)

10 + 1 * e([1]) + 5.2 * e([3,4]) + 3 * e([[2,3],4])


The OTI number ```a``` has been created. This number represents the following:

$$
a = 10 + \epsilon_1 + 5.2 \ \epsilon_3\epsilon_4+3 \ \epsilon_2^3\epsilon_4 
$$

therefore, the 'direction array' given as a parameter to the function ```oti.e(dirArray)``` correspond as follows:

- ```oti.e(i)``` :$\epsilon_i$ (notice it is just an  unsigned integer)

- ```oti.e([i,j,k,l])``` : $\epsilon_i\epsilon_j\epsilon_k\epsilon_l$ (Notice it's a one level array, with $n$ elements)

- ```oti.e([[i,m],j,[k,n],l])``` : $\epsilon_i^m\epsilon_j\epsilon_k^n\epsilon_l$ (Notice it has some positions where there are other arrays. Those arrays indicate the exponent of the number. Usually $m$, $n$ are greater than 1.)


To visualize the number, it can be done in two ways:

- A compact representation, which will show the constructor of the class:

In [49]:
a

10 + 1 * e([1]) + 5.2 * e([3,4]) + 3 * e([[2,3],4])

- Using string conversion, or just the print command. This will show a human readable version of the number.

In [50]:
print(a)

10 + 1 * e([1]) + 5.2 * e([3,4]) + 3 * e([[2,3],4])


The number has been created with the maximum truncation order corresponding to the maximum order of the given term. To know the order of the number, just call the atribute 'order'

In [6]:
print(a.order)

4


For more advanced users, there is a shorter representation of the number. To access it call the method ```short_repr()``` as indicated below. It shows only the real part. the total number of non-zero imaginary directions and the truncation order of the number.

In [7]:
print(a.short_repr())

sotinum(10.0, nnz: 4, order: 4)


There is also a long representation available, accessible through the call ```long_repr()```, as shown below:

In [9]:
print(a.long_repr())

sotinum(10.0, nnz: 4, alloc: 4, actual order: 4, truncation order: 4, flag: 1
  - Order 1->   nnz: 1  size: 1 
  - Order 2->   nnz: 1  size: 1 
  - Order 3->   nnz: 0  size: 0 
  - Order 4->   nnz: 1  size: 1 
)


#### 1.1.2 Display options
Each OTI number is displayed with a maximum of 4 imaginary directions and with format "%g". 

The current version of the library supports changing the representation using the ```set_printoptions``` function as follows:

```
oti.set_printoptions(float_format="g",terms_print=4)
```
- ```float_format:``` sets the format to print the coefficients in the OTI number.
- ```terms_print:``` sets the number of terms to print from an OTI number.

In [10]:
help(oti.set_printoptions)

Help on built-in function set_printoptions in module pyoti.sparse:

set_printoptions(...)
    PURPOSE:   Set the print options for OTI numbers.
    
    INPUTS:
        - float_format: (Default "g") Define the print format for t
        - terms_print: (Default 4) Number of terms to print.
    
    In order to reset the values to defaults, just call set_printoptions() with no arguments.



### 1.2 Math

In ```pyoti```, OTI numbers have overloaded operators so that operations work as similar as other scalar types in Python, like ```float``` or ```int```. The followign provides a summary of operations supported in ```pyoti```:

| Operation | Symbol | 
|---:|:---:|
| Addition | ```a+b``` |
| Subtraction | ```a-b``` |
| Multiplication | ```a*b``` |
| Division | ```a/b``` |
| Power | ```a**b``` |
| Subtraction | ```-``` |


#### 1.2.1. Arithmetic operations

--> **Addition**

To add two OTIs simply use the ```+``` operator:

For example, adding the numbers 

$$
a = 10 + \phantom{2.5 } \epsilon_1 + 5.2 \ \epsilon_3\epsilon_4+3 \ \epsilon_2^3\epsilon_4
$$

$$
b = 10 + 2.5 \ \epsilon_1 -5.2 \ \epsilon_3^3\epsilon_4 \phantom{+3 \ \epsilon_2^3\epsilon_4}
$$

$$
a+b = 20 + 3.5  \ \epsilon_1 +0 \ \epsilon_3\epsilon_4 + 3 \ \epsilon_2^3\epsilon_4
$$

In [51]:
a = 10.0 +     oti.e([1]) + 3.0*oti.e([[2,3],4]) + 5.2*oti.e([3,4]) 
b = 10.0 + 2.5*oti.e([1])                        - 5.2*oti.e([3,4])

sum_result = a+b
print('The result is:', sum_result)
print("With truncation order:", sum_result.order)

The result is: 20 + 3.5 * e([1]) + 0 * e([3,4]) + 3 * e([[2,3],4])
With truncation order: 4


Notice that the result has order 4, despite adding two numbers with different orders ($O($ ```a``` $) = 4$, $O($ ```b``` $) = 2$). **Addition preserves the maximum truncation order**.

--> **Multiplication**

In order to Multiply two OTI numbers in the library, just use the ```*``` operator:

In [24]:
mult_result = a*b
print("The result of a*b:",mult_result)
print("Order of a*b:     ",mult_result.order)

The result of a*b: 100 + 35 * e([1]) + 2.5 * e([[1,2]]) + 0 * e([3,4]) + ... 
Order of a*b:      4


Notice that the order of the result is 4. Therefore, for the multiplication, the **maximum truncation order** is preserved.



--> **Subtraction**

To subtract two OTI numbers, use ```-``` operator:

In [25]:
sub_result = a-b

print("The result of a-b:",sub_result)
print("Order of a-b:     ",sub_result.order)

The result of a-b:  - 1.5 * e([1]) + 10.4 * e([3,4]) + 3 * e([[2,3],4])
Order of a-b:      4


--> **Division**

Division between two OTIs is performed as follows:

In [26]:
div_result = a/b
print("The result of a/b:",div_result)
print("Order of a/b:     ",div_result.order)

The result of a/b: 1 - 0.15 * e([1]) + 0.0375 * e([[1,2]]) + 1.04 * e([3,4]) + ... 
Order of a/b:      4


#### --> Elementary functions.

```pyoti``` supports multiple functions, such as sin, cos, asin, acos, tan, atan, log(natural log), log10, exp, cosh, sinh, etc.

In [29]:
x = 3 + 2*oti.e([1]) - 4.3 * oti.e([2,3])
f = oti.sin(x)
print(f)

0.14112 - 1.97998 * e([1]) - 0.28224 * e([[1,2]]) + 4.25697 * e([2,3])


### 1.3 Get and set coefficients of an OTI number.


#### 1.3.1 Get coefficients of an OTI number.

To get a coefficient, just call the method '```get_im(dirArray)```' and place the direction array coefficient you want.


In [12]:
a = 10.0 + oti.e([1]) + 3.0*oti.e([[2,3],4]) + 5.2*oti.e([3,4]) 
print(a)

10 + 1 * e([1]) + 5.2 * e([3,4]) + 3 * e([[2,3],4])


To get the coefficient of the direction $\epsilon_1$ use:

In [13]:
print(a.get_im(1))

1.0


or

In [14]:
print(a.get_im([1]))

1.0


To get the coefficient of $\epsilon_2^3\epsilon_4$ use:

In [15]:
print(a.get_im([[2,3],4]))

3.0


or

In [16]:
print(a.get_im([2,2,2,4]))

3.0


Coefficients that are not contained within the number will return zero. For example, the coefficient along direction $\epsilon_1\epsilon_2^2$ is:

In [17]:
print(a.get_im([1,[2,2]]))

0.0


There are two ways to get the real coefficient: Get the real 

In [52]:
print(a.get_im(0))

10.0


In [53]:
a.real

10.0

#### 1.3.2 Set coefficients of an OTI number 

Similar to the methods of getting coefficients, setting the values of coefficients works in the same way. If the coefficient already exist, it will be modified, and if the coefficient does not exist, it will be created. Use the method '```set_im(newValue,dirArray)```' to do so.


In [20]:
a = 10.0 + oti.e([1]) + 3.0*oti.e([[2,3],4]) + 5.2*oti.e([3,4]) 
print(a)

10 + 1 * e([1]) + 5.2 * e([3,4]) + 3 * e([[2,3],4])


In [21]:
a.set_im(7.3,[3,4])
print(a)

10 + 1 * e([1]) + 7.3 * e([3,4]) + 3 * e([[2,3],4])


In [22]:
a.set_im(4.2,[1,[3,2]])
print(a)

10 + 1 * e([1]) + 7.3 * e([3,4]) + 4.2 * e([1,[3,2]]) + ... 


## 3. Applications

In general, dual numbers have most applications into finding derivatives of functions in an exact way. In this sense let's see some applications, different approaches to get the results and some comparisons with other methods such as multicomplex numbers.

### 3.1 Derivatives of polynomials
Dual numbers in general can be used to find derivatives of functions (whichever they are). The most strightforward way to use them is into polynomials.

#### 3.1.1 Single variable functions

Consider a $C^n$ function $f: \mathbb{R} \rightarrow \mathbb{R}$. Taking into account the Taylor series expansion of any function

$$
f(z) = \sum_{k=0}^{\infty}    \frac{f^k(a)}{k!}(z-a)^k
$$

Let´s take $z=x_0+\epsilon_1$ and $a=x_0$. Then what we get is the following:

$$
f(x_0+\epsilon) = f(x_0) + f_x(x_0)\epsilon_1 +
\frac{f_{xx}(x_0)}{2!}{\epsilon_1^2} +
\frac{f_{xxx}(x_0)}{3!}{\epsilon_1^3} + 
\frac{f_{xxxx}(x_0)}{4!}{\epsilon_1^4} + ...
$$

Notice that we get at each $i$'th order direction the $i$'th order derivative evaluated at $a=x_0$. 

To apply this in a real example let's consider the following function:

$$
f(x)= 10x^5-3x^3+2x-10
$$

In [54]:
def funct_1(x):
    return 10.0*x**5 -3.0*x**3+2.0*x-10.0

The analytical defivatives of the funciton are:
$$
f_{x}(x)= 50x^4-9x^2+2, \\ 
f_{xx}(x)= 200x^3-18x,  \\ 
f_{xxx}(x)= 600x^2-18, \\ 
f_{xxxx}(x)= 1200x, \\ 
f_{xxxxx}(x)= 1200,
$$

If we consider an $x_0=2.5$, then each derivative we obtain becomes:
$$
f(2.5) = 924.6875 \\
f_{x}(2.5)= 1898.875, \\ 
f_{xx}(2.5)= 3080,  \\ 
f_{xxx}(2.5)= 3732, \\ 
f_{xxxx}(2.5)= 3000, \\ 
f_{xxxxx}(2.5)= 1200,
$$

Let's now apply the taylor series expansion to the number. Therefore let's define $x_e = 2.5+\epsilon_1$. Since we want to obtain higher order terms, we need to redefine the order of the number. Therefore we will redefine it to order 5 (because we want to get  up to the fifth derivative).

In [56]:
x_e=2.5+oti.e([1],order=5)

print(x_e)
print(x_e.order)

2.5 + 1 * e([1])
5


Now let's evaluate the function in $x_e$

In [57]:
result_1=funct_1(x_e)
print(result_1)
result_1

924.688 + 1898.88 * e([1]) + 1540 * e([[1,2]]) + 622 * e([[1,3]]) + ... 


924.688 + 1898.88 * e([1]) + 1540 * e([[1,2]]) + 622 * e([[1,3]]) + ... 

We see that the first two values correspond to the function and first derivative evaluated at $2.5$, but let's see what happens with the numbers. In order to retreive the values of the other derivatives, the solution must take into account the following:

if our number results in something like
$$
\mbox{result} = a_0 + a_1\epsilon_1 + a_2\epsilon_1^2 + a_3\epsilon_1^3 + a_4\epsilon_1^4+... 
$$

and considering that 
$$
f(x_0+\epsilon) = f(x_0) + f_x(x_0)\epsilon_1 +
\frac{f_{xx}(x_0)}{2!}{\epsilon_1^2} +
\frac{f_{xxx}(x_0)}{3!}{\epsilon_1^3} + 
\frac{f_{xxxx}(x_0)}{4!}{\epsilon_1^4} + ...
$$

then to retreive the functions, we most do the following:
$$
f(x_0) = a_0 \\
f_x(x_0) = a_1 \\
f_{xx}(x_0)=a_2*2! \\
f_{xxx}(x_0)=a_3*3! \\ 
f_{xxxx}(x_0)=a_4*4! \\ 
\mbox{and so on...}
$$

As a result, in order to evaluate the result, consider the following:

In [58]:
f=result_1.get_im([0])

f_x=result_1.get_im([1])

f_xx=result_1.get_im([[1,2]]) * 2*1

f_xxx=result_1.get_im([[1,3]]) * 3*2*1

f_xxxx=result_1.get_im([[1,4]]) * 4*3*2*1

f_xxxxx=result_1.get_im([[1,5]]) * 5*4*3*2*1

# Print results:

print("f:")
print(f)

print("\nf_x:")
print(f_x)

print("\nf_xx:")
print(f_xx)

print("\nf_xxx:")
print(f_xxx)

print("\nf_xxxx:")
print(f_xxxx)

print("\nf_xxxxx:")
print(f_xxxxx)

f:
924.6875

f_x:
1898.875

f_xx:
3080.0

f_xxx:
3732.0

f_xxxx:
3000.0

f_xxxxx:
1200.0


An easier way to do this is to use the method ```get_deriv()``` which applies directly the factor to obtain the derivative.

In [63]:
print("f_xxx:")
print(result_1.get_deriv([1,1,1]))

f_xxx:
3732.0


Which for the sake of comparison, let's recall the results:

$$
f(2.5) = 924.6875 \\
f_{x}(2.5)= 1898.875, \\ 
f_{xx}(2.5)= 3080,  \\ 
f_{xxx}(2.5)= 3732, \\ 
f_{xxxx}(2.5)= 3000, \\ 
f_{xxxxx}(2.5)= 1200
$$

We see exact computation of the function's derivatives.

#### 3.1.2 Multi variable functions

Consider a function $f: \mathbb{R}^n \rightarrow \mathbb{R}$,  $f(x_1,x_2,...,x_n)$. Taking into account the Taylor series expansion of multivariable functions

$$
f(z_1,z_2,...,z_n) = \\
\sum_{k_1=0}^{\infty}\sum_{k_2=0}^{\infty}...\sum_{k_n=0}^{\infty}    
\frac{(z_1-a_1)^{k_1}(z_2-a_2)^{k_2}...(z_n-a_n)^{k_n}}{k_1!k_2!...k_n!}
\left(
\frac{ \partial^{k_1+k_2+...+k_n}f }{ \partial x_1^{k_1} \partial x_2^{k_2} ... \partial x_n^{k_n} }
\right)
(a_1,a_2,...,a_n)
$$


Now consider $z_1=x_{1_0}+\epsilon_1$, $z_2=x_{2_0}+\epsilon_2$, ..., $z_n=x_{n_0}+\epsilon_n$; and $a_1 = x_{1_0}$, $a_2 = x_{2_0}$, ...,$a_n = x_{n_0}$, we obtain:

$$
f(x_{1_0}+\epsilon_1,x_{2_0}+\epsilon_2,...,x_{n_0}+\epsilon_n) = \\
\sum_{k_1=0}^{\infty}\sum_{k_2=0}^{\infty}...\sum_{k_n=0}^{\infty}    
\frac{\epsilon_1^{k_1}\epsilon_2^{k_2}...\epsilon_n^{k_n}}{k_1!k_2!...k_n!}
\left(
\frac{ \partial^{k_1+k_2+...+k_n}f }{ \partial x_1^{k_1} \partial x_2^{k_2} ... \partial x_n^{k_n} }
\right)
(x_{1_0},x_{2_0},...,x_{n_0})
$$


Notice again that the order of the derivative is the same order of the dual number that we get, i.e., the sum $k_1+k_2+...+k_n$ determines the order of the dual direction and the order of the derivative to associate.

For example, a two variable function up to order 2 derivatives gives:

$$
f(x_{1_0}+\epsilon_1,x_{2_0}+\epsilon_2) = \\
f(x_{1_0},x_{2_0}) + \\
\frac{\partial f }{ \partial x_1} \ \epsilon_1 + 
\frac{1}{2!}\frac{\partial^2 f }{ \partial x_1^2}(x_{1_0},x_{2_0})  \ \epsilon_1^2 +\\
\frac{\partial f }{ \partial x_2}(x_{1_0},x_{2_0})  \ \epsilon_2 + 
\frac{\partial^2 f }{ \partial x_1\partial x_2}(x_{1_0},x_{2_0})  \ \epsilon_2\epsilon_1 +
\frac{1}{2!}\frac{\partial^2 f }{ \partial x_2^2}(x_{1_0},x_{2_0}) \  \epsilon_2^2
$$

To apply this to a simple example, let's consider the function 

$$
f(x,y) = 4x^5y^4
$$

Let's define the function in the python environment:

In [64]:
def funct_2(x,y):
    return 4*x**5*y**4

Analitically, let's define the derivatives of the function:
$$
f_x(x,y) = 20x^4y^4, \ \ f_y(x,y) = 16x^5y^3
$$

Now, how do we retreive the function's first derivatives by just evaluating the function with dual numbers? It's easy! 

According to the taylor series expansion of the function, we can retreive the function derivatives by evaluating the function at the point we want, say $(x_0,y_0)$ with a perturbation on the dual direction. In this case, the perturbation will occur as follows: $(x_0+\epsilon_1,y_0+\epsilon_2)$

To get first derivatives we most only evaluate the function into dual numbers of order one. In this way, we define the point $(x_0=2,y_0=3)$ and we get:

In [66]:
x=2.0+oti.e([1])
y=3.0+oti.e([2])

In [67]:
result_2 = funct_2(x,y)
print(result_2)

10368 + 25920 * e([1]) + 13824 * e([2])


Analitically, the results of evaluating all function and its derivatives are:
$$
f(2,3) = 10368
$$
$$
f_x(2,3) = 25920
$$
$$
f_y(2,3) = 13824
$$

Now, interestingly, the coefficients that we obtain are the following:
``` 2048.0 ``` on the real part, wich represents the function evaluated at $(2,3)$, the coefficient of $\epsilon_1$ is the first derivative with respect to $x$ (as is) and the coefficient of $\epsilon_2$ is the first derivative with respect to $y$ function evaluated.

The seccond derivatives of the function are:
$$
f_{xx}(x,y) = 80x^3y^4, \\
f_{xy}(x,y) = 80x^4y^3, \\
f_{yy}(x,y) = 48x^5y^2
$$

Wich evaluated at $(2,3)$ gives
$$
f_{xx}(2,3) = 51840, \\ 
f_{xy}(2,3) = 34560, \\
f_{yy}(2,3) = 13824
$$


Lets now change the order of both numbers, and see what we get:


In [44]:
# To change the order on the fly, perturb by zero-coefficient OTI numbers with desired truncation order.
x = x + 0*oti.e([1],order=2) 
y = y + 0*oti.e([1],order=2) 
result_2_order2 = funct_2(x,y)
print(result_2_order2)

10368 + 25920 * e([1]) + 13824 * e([2]) + 25920 * e([[1,2]]) + ... 


We can extract, based on the Taylor series, the different terms. However, in order to get an easy form to extract them, the function ```getDerivative(dualnum,dirArray)``` helps in retreiving the derivatives according to the specifyed direction. It works as follows:

In [45]:
print("Get the f(x_0, y_0):")
print(result_2_order2.get_deriv([0]))

print("\nGet the f_x(x_0, y_0):")
print(result_2_order2.get_deriv([1]))

print("\nGet the f_xx(x_0, y_0):")
print(result_2_order2.get_deriv([[1,2]]))

print("\nGet the f_y(x_0, y_0):")
print(result_2_order2.get_deriv([2]))

print("\nGet the f_xy(x_0, y_0):")
print(result_2_order2.get_deriv([1,2]))

print("\nGet the f_yy(x_0, y_0):")
print(result_2_order2.get_deriv([[2,2]]))

Get the f(x_0, y_0):
10368.0

Get the f_x(x_0, y_0):
25920.0

Get the f_xx(x_0, y_0):
51840.0

Get the f_y(x_0, y_0):
13824.0

Get the f_xy(x_0, y_0):
34560.0

Get the f_yy(x_0, y_0):
13824.0


This finalizes the method to obtain the derivatives of polynomial functions.

## 4 Linear Algebra

```pyoti``` includes support for linear algebra operations (such as matrix multiplication, dot product, etc).